### PR TITLE
[New] Add ENC exchange set

### PR DESCRIPTION
**Pull request by chriswebb09**
## Description

This PR implements **`Add ENC Exchange Set`** in `Layers` category.

`swift/issues/5512`

## How To Test

Open view and wait for ENC data to download, once loaded it should load onto the map and render.

## Screenshots

add-enc-exchange-set

### CODE REVIEW

File: Shared/Samples/Add ENC Exchange Set/AddENCExchangeSetView.swift [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+            let result = exchangeSet.loadStatus
+            if result == .loaded {
+                try await update(
+                    dataSet: exchangeSet.datasets,
+                    mapProxy: proxy
+                )
+            }
+        }
+        
+        private func update(dataSet: [ENCDataset], mapProxy: MapViewProxy) async throws {
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC Exchange Set/AddENCExchangeSetView.swift:106]: Rename function to something more descriptive

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+import ArcGIS
+import SwiftUI
+
+struct AddENCExchangeSetView: View {
+    /// The current draw status of the map.
+    @State private var currentDrawStatus: DrawStatus = .inProgress
+    
+    /// The error shown in the error alert.
+    @State var error: Error?
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:23]: make this state var private

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            let map = Map(basemapStyle: .arcGISOceans)
+            map.initialViewpoint = Viewpoint(
+                latitude: -32.5,
+                longitude: 60.95,
+                scale: 67000
+            )
+            return map
+        }()
+        
+        var completeExtent: Envelope?
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:71]: make private

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        
+        /// A URL to the temporary SENC data directory.
+        let temporaryURL: URL = {
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:74]: make private

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+            let resourceURL = URL.hydrographyDirectory.deletingLastPathComponent()
+            // Set environment settings for loading the dataset.
+            let environmentSettings = ENCEnvironmentSettings.shared
+            environmentSettings.resourceURL = resourceURL
+            // The SENC data directory is for temporarily storing generated files.
+            environmentSettings.sencDataURL = temporaryURL
+            updateDisplaySettings()
+            try await exchangeSet.load()
+            let result = exchangeSet.loadStatus
+            if result == .loaded {
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:98]: check load status directly from exchange set

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            )
+            // Create and return the full, unique URL to the temporary folder.
+            try? FileManager.default.createDirectory(
+                at: directoryURL,
+                withIntermediateDirectories: true
+            )
+            return directoryURL
+        }()
+        
+        func addENCExchangeSet(proxy: MapViewProxy) async throws {
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:86]: add docstring

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+            let result = exchangeSet.loadStatus
+            if result == .loaded {
+                try await renderENCData(
+                    dataSet: exchangeSet.datasets,
+                    mapProxy: proxy
+                )
+            }
+        }
+        
+        private func renderENCData(dataSet: [ENCDataset], mapProxy: MapViewProxy) async throws {
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:106]: add docstring

File: Samples.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -1688,14 +1694,29 @@
...
+		};
 		954AEDEF2C01332F00265114 /* Select features in scene layer */ = {
 			isa = PBXGroup;
 			children = (
 				954AEDED2C01332600265114 /* SelectFeaturesInSceneLayerView.swift */,
 			);
 			path = "Select features in scene layer";
-			sourceTree = "<group>";
-		};
+    };
```
Comment (chriswebb09) [inline Samples.xcodeproj/project.pbxproj:1719]: Fix this

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+                        self.error = error
+                    }
+                }
+        }
+        .errorAlert(presentingError: $error)
+    }
+}
+
+private extension AddENCExchangeSetView {
+    @MainActor
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:62]: I don't think this is neeeded

**Comment by chriswebb09** [comment]
@CalebRas @dfeinzimer this should run as expected now

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+            environmentSettings.resourceURL = resourceURL
+            // The SENC data directory is for temporarily storing generated files.
+            environmentSettings.sencDataURL = temporaryURL
+            updateDisplaySettings()
+            try await exchangeSet.load()
+            if exchangeSet.loadStatus == .loaded {
+                try await renderENCData(dataSet: exchangeSet.datasets)
+                if let extent = completeExtent {
+                    await mapProxy.setViewpoint(
+                        Viewpoint(center: extent.center, scale: 67000)
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:118]: check if it is possible to move this back to the view

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+    @State private var error: Error?
+    
+    /// The data model for the sample.
+    @StateObject private var model = Model()
+    
+    var body: some View {
+        MapViewReader { mapProxy in
+            MapView(map: model.map)
+                .onDrawStatusChanged { drawStatus in
+                    // Updates the state when the map's draw status changes.
```
Comment (CalebRas) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:32]: I am not sure this is necessary. Generally, we use a `ProgressView` to indicate on-going async operations, not for rendering. I think removing this `onDrawStatusChanged` modifier and simply setting `isLoading = false` after calling `addENCExchangeSet` is sufficient. Up to you, though.

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+                    await mapProxy.setViewpoint(
+                        Viewpoint(center: extent.center, scale: 67000)
+                    )
+                }
+            }
+        }
+        
+        /// Maps the exchange set data to ENC layer and ENC cells and loads the layers.
+        /// - Parameter dataSet: The ENC dataset previously loaded.
+        private func renderENCData(dataSet: [ENCDataset]) async throws {
```
Comment (CalebRas) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:126]:
```suggestion
        /// Maps the exchange set data to ENC layer and ENC cells and loads the layers.
        /// - Parameter dataset: The ENC dataset previously loaded.
        private func renderENCData(dataset: [ENCDataset]) async throws {
```

Dataset is one word.

File: Shared/Samples/Add ENC exchange set/README.md [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+## How it works
+
+1. Specify the path to a local CATALOG.031 file to create an `EncExchangeSet`.
+2. After loading the exchange set, get the `EncDataset` objects asynchronously.
+3. Create an `EncCell` for each dataset. Then create an `EncLayer` for each cell.
+4. Add the ENC layer to a map's operational layers collection to display it.
+
+## Relevant API
+
+* EncCell
```
Comment (CalebRas) [inline Shared/Samples/Add ENC exchange set/README.md:26]: "ENC" in the API names should be capitalized. For example, [ENCCell](https://developers.arcgis.com/swift/api-reference/documentation/arcgis/enccell). Can you updated them throughout the README and metadata please?

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 14]
```diff
@@ -0,0 +1,181 @@
...
+            let encLayers = dataSet.map {
+                ENCLayer(
+                    cell: ENCCell(
+                        dataset: $0
+                    )
+                )
+            }
+            for encLayer in encLayers {
```
[the remark this answers is not in the document]
Comment (chriswebb09) [reply Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift]: Works great!

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift (hunk 11 above)
Comment (chriswebb09) [reply Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift]: Yeah, you're right, it doesn't need that logic anymore.

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+            // The SENC data directory is for temporarily storing generated files.
+            environmentSettings.sencDataURL = temporaryURL
+            updateDisplaySettings()
+            try await exchangeSet.load()
+            try await renderENCData(dataset: exchangeSet.datasets)
+        }
+        
+        /// Maps the exchange set data to ENC layer and ENC cells and loads the layers.
+        /// - Parameter dataset: The ENC dataset previously loaded.
+        private func renderENCData(dataset: [ENCDataset]) async throws {
```
Comment (CalebRas) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:116]:
```suggestion
        /// - Parameter datasets: The ENC datasets previously loaded.
        private func renderENCData(datasets: [ENCDataset]) async throws {
```

Sorry I didn't realize this was an array the first time around.

**Review by CalebRas — COMMENTED** [review]
Looks good! Just one more thing.

File: Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+            environmentSettings.sencDataURL = temporaryURL
+            updateDisplaySettings()
+            try await exchangeSet.load()
+            try await renderENCData(dataset: exchangeSet.datasets)
+        }
+        
+        /// Maps the exchange set data to ENC layer and ENC cells and loads the layers.
+        /// - Parameter datasets: The ENC datasets previously loaded.
+        private func renderENCData(datasets: [ENCDataset]) async throws {
+            let encLayers = dataset.map {
```
Comment (chriswebb09) [inline Shared/Samples/Add ENC exchange set/AddENCExchangeSetView.swift:117]: fix

**Review by CalebRas — DISMISSED** [review]

**Comment by chriswebb09** [comment]
Just added one more sentence to README. 
@dfeinzimer @CalebRas

File: Shared/Samples/Add ENC exchange set/README.md [hunk 17]
```diff
@@ -30,7 +30,9 @@ Run the sample and view the ENC data. Pan and zoom around the map. Take note of
 
 ## Offline data
 
-This sample downloads the [ENC Exchange Set without updates](https://www.arcgis.com/home/item.html?id=9d2987a825c646468b3ce7512fb76e2d) and [Hydrography dataset resources](https://www.arcgis.com/home/item.html?id=c16f3845b9cc4b93a908d87d28823afd) items from ArcGIS Online.
+This sample downloads the [ENC Exchange Set without updates](https://www.arcgis.com/home/item.html?id=9d2987a825c646468b3ce7512fb76e2d) item from ArcGIS Online.
+
+The latest [hydrography package](https://www.arcgis.com/home/item.html?id=c16f3845b9cc4b93a908d87d28823afd) can be downloaded from ArcGIS for Developers. The S57DataDictionary.xml file is contained in it along with many others but a user does not need to know that in order to render ENC data.
```
Comment (dfeinzimer) [inline Shared/Samples/Add ENC exchange set/README.md:35]: I don't think we normally deviate from the common design in the READMEs. @yo1995 are we okay to deviate here?

Comment (yo1995) [reply Shared/Samples/Add ENC exchange set/README.md:35]: This is an intended change. We use to have it in the iOS README: https://github.com/Esri/arcgis-runtime-samples-ios/tree/main/arcgis-ios-sdk-samples/Layers/Add%20ENC%20exchange%20set#offline-data

However, the URL should link to the download page, not the AGOL portal item. The purpose is to let a developer know where to find the source-of-truth hydro dataset.

**Review by dfeinzimer — APPROVED** [review]

**Review by CalebRas — APPROVED** [review]

**Review by yo1995 — APPROVED** [review]

**Pull request merged**